Automattic/woocommerce-payments
# Add support for (custom) order numbers in API calls

Fixes #4142

#### Changes proposed in this Pull Request

WooCommerce Orders have IDs, which are the same as their post ID, and order numbers, which by default are also the same, but can be customized through extensions. Those are retrieved through `$order->get_id()` and `$order->get_order_number()` respectively.

Within WCPay we seem to use `get_order_number` everywhere within the site, but when we're sending information to Stripe, we only included the order/post ID. **This PR** adds the number to the metadata we're sending to Stripe, and updates the description of the payment to use the order number instead of ID.

Here's the description (with this PR):

<img width="830" alt="image" src="https://user-images.githubusercontent.com/5311119/169290515-41c75fc3-2bb1-451c-889c-df90aef4317b.png">
and the updated metadata:
<img width="514" alt="Screenshot 2022-05-19 at 15 12 47" src="https://user-images.githubusercontent.com/5311119/169290704-74c4b771-dfe0-459b-933e-b3dd980e9c5f.png">

This was done, implemented, and manually tested in the following scenarios:

- UPE standard card
- UPE SCA card
- UPE redirect
- classic checkout
- express checkout

#### Testing instructions

0. Implement a manual filter, or setup an extension like [Sequential Order Number for WooCommerce](https://wordpress.org/plugins/wt-woocommerce-sequential-order-numbers/). This way you'll be able to see the difference between IDs and numbers, otherwise they would be the same.
1. Make payments in various ways (the ones listed above, and any others if you think of them)
2. Check the payments in Stripe's dashboard, and make sure that the metadata includes both order ID and number, and that the description uses the number.


-------------------
